// corespring-match/src/corespring-match.jsx
import React, { PropTypes } from 'react';
import CorespringCorrectAnswerToggle from 'corespring-correct-answer-toggle';

import update from 'immutability-helper';
import * as _ from 'lodash';
import ChoiceInput from 'choice-input';
import NothingSubmittedIcon from 'corespring-icon/nothing-submitted-icon';
import FeedbackPanel from 'corespring-feedback-panel';

export default class CorespringMatch extends React.Component {
  
  constructor(props) {
    super(props);
    this.state = {
      showCorrect: false
    };
  }

  componentWillMount(props) {
    this.setState({
      model: this._prepareModel() 
    });
  }

  componentWillReceiveProps(props) {
    this.setState({
      model: this._prepareModel(props) 
    });
  }

  _className() {
    return 'corespring-match-table';
  }

  _prepareModel(props) {
    const $log = console;
    const YES_LABEL = 'Yes';
    const YES_NO = 'YES_NO';
    const TRUE_FALSE = 'TRUE_FALSE';
    const TRUE_LABEL = 'True';
    const FALSE_LABEL = 'False';

    let self = this;
    props = props !== undefined ? props : this.props;

    let prepareColumns = () => {
      let columns = _.cloneDeep(props.model.columns);
      let answerType = props.model.answerType;
      if (answerType === YES_NO || answerType === TRUE_FALSE) {
        if (columns.length !== 3) {
          $log.warn('Match interaction with boolean answer type should have 2 columns, found ' + columns.length);
          while (columns.length < 3) {
            columns.push({
              labelHtml: ''
            });
          }
        }
        if (_.isEmpty(columns[1].labelHtml)) {
          columns[1].labelHtml = answerType === TRUE_FALSE ? TRUE_LABEL : YES_LABEL;
        }
        if (_.isEmpty(columns[2].labelHtml)) {
          columns[2].labelHtml = answerType === TRUE_FALSE ? FALSE_LABEL : NO_LABEL;
        }
      }

      function isDefaultLabel(s){
        switch(s){
          case "Custom header":
          case "Column 1":
          case "Column 2":
          case "Column 3":
          case "Column 4":
          case "Column 5":
            return true;
          default:
            return false;
        }
      }

      _.forEach(columns, function(col, index) {
        col.cssClass = index === 0 ? 'question-header' : 'answer-header';
        col.labelHtml = isDefaultLabel(col.labelHtml) ? '' : col.labelHtml;
      });

      return columns;
    };


    let whereIdIsEqual = (id) => {
      return function(match) {
        return match.id === id;
      };
    };

    let prepareRows = () => {
      let createMatchSetFromSession = (id) => {
        return _.find(props.session.answers, whereIdIsEqual(id))
          .matchSet.map((match) => {
            return {
              value: match
            };
          });
      };

      let createEmptyMatchSet = (length) => {
        return _.range(length).map(() => {
          return {
            value: false
          };
        });
      };

      let answersExist = (props.session && props.session.answers);
      let rows = props.model.rows.map((row) => {
        let cloneRow = _.cloneDeep(row);
        cloneRow.matchSet = answersExist === true ? createMatchSetFromSession(row.id) : createEmptyMatchSet(props.model.columns.length - 1);
        return cloneRow;
      });
      return rows;
    };

    return {
      columns: prepareColumns(),
      rows: prepareRows() //this.props.model.config.shuffle ? _.shuffle(prepareRows()) : prepareRows()
    };
  }

  change(rowIndex, columnIndex, value) {
    let stateToSession = (state) => {
      return state.model.rows.map(row => {
        return {
          id: row.id,
          matchSet: row.matchSet.map(match => match.value)
        };
      });
    };

    let callback = () => {
      let session = stateToSession(this.state);
      this.props.session.answers = session;
      this.props.onChange(session);
    };

    if (this.props.model.config.inputType === 'radio') {
      let row = Array.apply(null, Array(this.props.model.columns.length - 1)).map((a, index) => {return {value: (index === columnIndex) };});
      this.setState(update(this.state, { model: { rows: { [rowIndex] : { matchSet: { "$set": row } } } } }), callback);
    } else {
      this.setState(update(this.state, { model: { rows: { [rowIndex] : { matchSet: { [columnIndex] : { value: { "$set": value } } } } } } }), callback);      
    }
  }

  onToggle() {
    if (this.props.mode === 'evaluate') {
      this.setState(update(this.state, {showCorrect: {"$set" : !this.state.showCorrect}}));
    }
  }

  render() {
    let self = this;
    let rows = this.state.model.rows;
    let columns = this.state.model.columns;
    let disabled = this.props.mode !== 'gather';
    let showCorrect = (this.props.mode === 'evaluate' && this.state.showCorrect);

    let correctness = (rowIndex, columnIndex) => {
      if (showCorrect && this.props.model.correctResponse) {
        return this.props.model.correctResponse[rowIndex].matchSet[columnIndex] ? 'correct' : undefined;
      } else if (this.props.model.correctnessMatrix) {
        return this.props.model.correctnessMatrix[rowIndex].matchSet[columnIndex].correctness;
      }
      return undefined;
    };

    let choiceCellClass = (rowIndex, columnIndex) => {
      return _.without([
        'match-td-padded',
        correctness(rowIndex, columnIndex)
      ], undefined).join(' ');
    }

    let answerExpected = (rowIndex) => {
      return !showCorrect && (this.props.model.correctnessMatrix && this.props.model.correctnessMatrix[rowIndex].answerExpected);
    }

    let checked = (rowIndex, columnIndex) => {
      let correctChecked = !_.includes(['unknown', undefined], correctness(rowIndex, columnIndex))
      if (showCorrect) {
        return this.props.model.correctResponse[rowIndex].matchSet[columnIndex];
      } else {
        return correctChecked || rows[rowIndex].matchSet[columnIndex].value;
      }
    };

    let showToggle = this.props.mode === 'evaluate' && this.props.model.numAnswers !== 0 && this.props.model.correctness !== 'correct';

    return <div className="corespring-match">
      {
        (showToggle) ? <div className="correct-answer-toggle">
          <CorespringCorrectAnswerToggle
            show={showToggle}
            toggled={this.state.showCorrect}
            onToggle={this.onToggle.bind(this)} />
          </div> :
        <div/>
      }
      <table className={self._className()}>
        <thead>
          <tr className="header-row">
          {
            columns.map((column, index) => {
              return <th colSpan={index?1:2} key={index} dangerouslySetInnerHTML={{__html: column.labelHtml}}></th>;
            })
          }
          </tr>
        </thead>
        <tbody>
          { 
            rows.map((row, rowIndex) => {
              return <tr className={`question-row row-${row.id}`} key={rowIndex}>
                <td className="question-cell match-td-padded" dangerouslySetInnerHTML={{__html: row.labelHtml}}></td>
                <td className="answer-expected-warning match-td-padded">
                {
                  (answerExpected(rowIndex)) ?
                    <div className="warning-holder">
                      <NothingSubmittedIcon iconSet="check" category="feedback" shape="square" />
                    </div> :
                    <div/>
                }
                </td>
                {
                  row.matchSet.map((match, columnIndex) => {
                    return <td className={choiceCellClass(rowIndex, columnIndex)} key={columnIndex}>
                      <ChoiceInput
                          choiceMode={self.props.model.config.inputType}
                          checked={checked(rowIndex, columnIndex)}
                          disabled={disabled}
                          correctness={correctness(rowIndex, columnIndex)}
                          showFeedbackTick={false}
                          onChange={(result) => { self.change(rowIndex, columnIndex, result.selected); }}
                        />
                    </td>;
                  })
                }
              </tr>;
            })
          }
        </tbody>
      </table>
      {
        (this.state.showCorrect) ? <div/> : <FeedbackPanel feedback={this.props.model.feedback} correctness={this.props.model.correctness} />
      }
    </div>;
  }

}

CorespringMatch.propTypes = {
  mode: PropTypes.oneOf(['gather', 'view', 'evaluate']),
  model: PropTypes.object,
  outcomes: PropTypes.array,
  session: PropTypes.object,
  onChange: PropTypes.func
};

CorespringMatch.defaultProps = {
  session: {
    answers: {}
  }
};
